### DISCUSSION TITLE
Git Repo Relative Path

### DISCUSSION THREAD

**Discussion by GavinOsborn** [JanDeDobbeleer/oh-my-posh#2717] [Troubleshoot]
Hey folks,

When you're terminal session is within a git repo is it possible to evaluate the path relative to the root of the repo?
eg: if the root of my repo was `~/code/github/foo/` and I was currently in `~/code/github/foo/baz/bar` then the path evaluated would be `baz/bar`.

It feels like the [Mapped Locations sample](https://ohmyposh.dev/docs/segments/path#mapped-locations) was in response to a similar question - but I was hoping for something that was contextual to the current repo.

That in itself would be great. Making it available to the path formatting styles functions would be better still.

Thanks

**Comment by JanDeDobbeleer** [comment]
@GavinOsborn let me have a quick play around with it as I think this can be done using cross segment properties.

**Reply by JanDeDobbeleer** [reply]
@GavinOsborn when #2719 lands, you can do something like this:

```json
"template": "{{ if and .Segments.Git.RepoName (ne .Segments.Git.RepoName .Path) }} \uf1d2 {{ .Segments.Git.RepoName }} \uf554 {{ trimPrefix .Segments.Git.Dir .PWD }}{{ else }} \uf74a {{ path .Path .Location }}{{ end }}"
```

**Reply by GavinOsborn** [reply]
Wow, thanks @JanDeDobbeleer  for such a prompt response!

So the solution you posted works great unless you're in the root of the repo in which case it prints the full path which isn't quite what I was after. This happens because the `.Segments.Git.Dir` is suffixed with a trailing `\` but `.PWD` is not. My knowledge of Go templates is limited to what I could google but I resolved it by appending a `\` to `.PWD` before evaluation.

```json
"template": "{{ if and .Segments.Git.RepoName (ne .Segments.Git.RepoName .Path) }}{{ $repoRoot := .Segments.Git.Dir }}{{ $repoPath := list .PWD \"\\\\\" | join \"\"  }} \uF408 \uf554 {{ replace $repoRoot \"\" $repoPath  }}{{ else }}<#f2f200> \ue5ff </>{{ path .Path .Location }}{{ end }}"
```

I don't know oh-my-posh well enough to comment usefully yet but from this first contact experience I would suggest a couple of things:
 - Standardize how paths are returned from the various segments you support. 
   - eg: a directory paths is never (or always?) suffixed with a path separator.
 - Consider support for common literals like `/` and `\`
   - better still perhaps a value that correctly represents the path separator for the local OS?
   
But take my feedback with a pinch of salt - I'm still figuring my way around 😊 

Thanks again for your help!

**Reply by JanDeDobbeleer** [reply]
@GavinOsborn git is always using those paths, I'll trim it off. Expect a fix asap (probably tomorrow as it's 23:00 here).

**Comment by GavinOsborn** [comment]
Take your time. It's only software 😄

**Reply by JanDeDobbeleer** [reply]
I have it ready, will push ASAP ;-)

**Reply by matcls** [reply]
Hola,
Testing I noticed that it works on linux but not on windows.

On Windows
![GitRoot](https://user-images.githubusercontent.com/60364655/187805721-5bc6aaf8-d73b-4d34-81a2-6981d0881582.png)
these are the values for the properties that I get 
```
C:\Users\user\rep_test\dir
.RepoName::/Users/user/rep_test 
.Dir:~/rep_test 
.PWD:~\rep_test\dir 
```
And this is how it looks like on linux
![SR58LU~Z](https://user-images.githubusercontent.com/60364655/187807820-edc578d4-f2ee-419b-be40-4c7e69bb222c.PNG)

Thanks

**Reply by lewis-yeung** [reply]
> it works on linux but not on windows.

@JanDeDobbeleer Bug confirmed. We should fix it.

**Reply by JanDeDobbeleer** [reply]
@lewis-yeung ah yes, will have a look.

**Reply by lewis-yeung** [reply]
@JanDeDobbeleer I have got a fix already and I'm ready to create a PR. Before doing this, I'm just wondering if we could just call `filepath.Base` instead of [`environment.Base`](https://github.com/JanDeDobbeleer/oh-my-posh/blob/2a1fdffb73efce9ef2f70e4fb6fa73c29aef287f/src/environment/shell.go#L786-L813) to get the last element of a path.

**Reply by JanDeDobbeleer** [reply]
We could, but is there an issue with testing in this case? It's been a long while that function was implemented. You can try.

**Reply by lewis-yeung** [reply]
@JanDeDobbeleer No yet. I have just read through the code and found the differences between them, since the current implementation works fine, I won't change it. :)

**Reply by lewis-yeung** [reply]
@JanDeDobbeleer Oops. We know that Windows accepts `/` as path separator. Now I know why `.RepoName` was populated with `:/Users/user/rep_test` (mentioned [here](https://github.com/JanDeDobbeleer/oh-my-posh/discussions/2717#discussioncomment-3522468)). I believe we need to adjust these:

https://github.com/JanDeDobbeleer/oh-my-posh/blob/2a1fdffb73efce9ef2f70e4fb6fa73c29aef287f/src/environment/shell.go#L792

https://github.com/JanDeDobbeleer/oh-my-posh/blob/2a1fdffb73efce9ef2f70e4fb6fa73c29aef287f/src/environment/shell.go#L802

just like that in `filepath.Base`, we should ~~use `os.IsPathSeparator` instead~~ implement a simple function like `os.IsPathSeparator`.

**Reply by lewis-yeung** [reply]
@matcls The bug has been fixed. Please try the latest release.

**Reply by matcls** [reply]
> @matcls The bug has been fixed. Please try the latest release.

Awesome, thanks!
